tests/optimize/curve_fit_test.py

```diff
@@ -0,0 +1,119 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2022 Scipp contributors (https://github.com/scipp)
+import numpy as np
+import scipp as sc
+from scipp.optimize import curve_fit
+
+import pytest
+
+
+def func(x, a, b):
+    return a * sc.exp(-(b / x.unit) * x)
+
+
+def array1d(*, a=1.2, b=1.3, noise_scale=0.1):
+    size = 50
+    x = sc.linspace(dim='xx', start=-0.1, stop=4.0, num=size, unit='m')
+    y = func(x, a, b)
+    rng = np.random.default_rng()
+    # Noise is random but avoiding unbounded values to avoid flaky tests
+    y.values += noise_scale * np.clip(rng.normal(size=size), -2.0, 2.0)
+    return sc.DataArray(y, coords={'xx': x})
+
+
+def test_should_not_raise_given_function_with_dimensionless_params_and_1d_array():
+    curve_fit(func, array1d())
+
+
+def test_should_raise_TypeError_when_xdata_given_as_param():
+    with pytest.raises(TypeError):
+        curve_fit(func, array1d(), xdata=np.arange(4))
+
+
+def test_should_raise_TypeError_when_ydata_given_as_param():
+    with pytest.raises(TypeError):
+        curve_fit(func, array1d(), ydata=np.arange(4))
+
+
+def test_should_raise_TypeError_when_sigma_given_as_param():
+    with pytest.raises(TypeError):
+        curve_fit(func, array1d(), sigma=np.arange(4))
+
+
+def test_should_raise_NotFoundError_when_data_array_has_no_coord():
+    da = array1d()
+    del da.coords[da.dim]
+    with pytest.raises(sc.NotFoundError):
+        curve_fit(func, da)
+
+
+def test_should_raise_BinEdgeError_when_data_array_is_histogram():
+    da = array1d()
+    hist = da[1:].copy()
+    hist.coords[hist.dim] = da.coords[hist.dim]
+    with pytest.raises(sc.BinEdgeError):
+        curve_fit(func, hist)
+
+
+def test_masks_are_not_ignored():
+    da = array1d()
+    unmasked, _ = curve_fit(func, da)
+    da.masks['mask'] = sc.zeros(sizes=da.sizes, dtype=bool)
+    da.masks['mask'][0] = True
+    masked, _ = curve_fit(func, da)
+    assert all(masked != unmasked)
+
+
+@pytest.mark.parametrize("noise_scale", [1e-1, 1e-2, 1e-3, 1e-6, 1e-9])
+def test_optimized_params_approach_real_params_as_data_noise_decreases(noise_scale):
+    popt, _ = curve_fit(func, array1d(a=1.7, b=1.5, noise_scale=noise_scale))
+    np.testing.assert_allclose(popt, [1.7, 1.5], rtol=noise_scale)
+
+
+@pytest.mark.parametrize("mask_pos", [0, 1, -3])
+@pytest.mark.parametrize("mask_size", [1, 2])
+def test_masked_points_are_treated_as_if_they_were_removed(mask_pos, mask_size):
+    da = array1d()
+    da.masks['mask'] = sc.zeros(sizes=da.sizes, dtype=bool)
+    da.masks['mask'][mask_pos:mask_pos + mask_size] = sc.scalar(True)
```

Comment:
In order for this to work reliably, you should set the size of `da` via an arg of `array1d`.

Reply:
Changed.